File: app/src/main/java/com/salino/sali/data/service/ItemNameAutocompleteMatcher.kt
package com.salino.sali.data.service

import com.salino.sali.util.normalizeItemName

internal object ItemNameAutocompleteMatcher {

    fun prefixMatchScore(displayName: String, query: String): Int {
        val normalizedQuery = normalizeItemName(query)
        if (normalizedQuery.isBlank()) return 0

        val normalizedDisplay = normalizeItemName(displayName)
        if (normalizedDisplay.isBlank()) return 0

        if (normalizedDisplay.startsWith(normalizedQuery)) {
            return if (normalizedDisplay == normalizedQuery) 120 else 100
        }

        val firstToken = normalizedDisplay.split(" ").firstOrNull().orEmpty()
        return if (firstToken.startsWith(normalizedQuery)) 80 else 0
    }

    fun matchesPrefix(displayName: String, query: String): Boolean =
        prefixMatchScore(displayName, query) > 0
}

File: mobile/src/services/itemNameAutocompleteMatcher.ts
import { normalizeItemName } from '../utils/textUtils';

export function prefixMatchScore(displayName: string, query: string): number {
  const normalizedQuery = normalizeItemName(query);
  if (!normalizedQuery) return 0;

  const normalizedDisplay = normalizeItemName(displayName);
  if (!normalizedDisplay) return 0;

  if (normalizedDisplay.startsWith(normalizedQuery)) {
    return normalizedDisplay === normalizedQuery ? 120 : 100;
  }

  const firstToken = normalizedDisplay.split(' ')[0] ?? '';
  return firstToken.startsWith(normalizedQuery) ? 80 : 0;
}

export function matchesPrefix(displayName: string, query: string): boolean {
  return prefixMatchScore(displayName, query) > 0;
}
